slm_porosity/heat_transfer.f90

```diff
@@ -737,54 +660,108 @@ FUNCTION porosity_term ()
     porosity_term = (1.0_pr - u(:,n_var_porosity)) / (1.0_pr - initial_porosity)
   END FUNCTION porosity_term
 
-  PURE FUNCTION conductivity (temperature)
+  FUNCTION conductivity (enthalpy)
     IMPLICIT NONE
-    REAL (pr), INTENT(IN) :: temperature(:)
-    REAL (pr) :: conductivity(SIZE(temperature))
-    conductivity = 1.0_pr + conductivity_der*temperature
+    REAL (pr), INTENT(IN) ::  enthalpy(:)
+    REAL (pr) :: conductivity(SIZE(enthalpy))
+    conductivity = 1.0_pr + conductivity_der_solid*temperature(enthalpy) + &
+      (conductivity_fusion + (conductivity_der_liquid - conductivity_der_solid)*(temperature(enthalpy) - 1.0_pr)) &
+      *liquid_fraction(enthalpy)
   END FUNCTION conductivity
 
-  PURE FUNCTION capacity (temperature)
+  FUNCTION capacity (enthalpy)
     IMPLICIT NONE
-    REAL (pr), INTENT(IN) :: temperature(:)
-    REAL (pr) :: capacity(SIZE(temperature))
-    capacity = 1.0_pr + 2*capacity_der*temperature
+    REAL (pr), INTENT(IN) :: enthalpy(:)
+    REAL (pr) :: capacity(SIZE(enthalpy))
+    capacity = 1.0_pr + capacity_der_solid*temperature(enthalpy) + &
+      (capacity_fusion + (capacity_der_liquid - capacity_der_solid)*(temperature(enthalpy) - 1.0_pr)) &
+      *liquid_fraction(enthalpy)
   END FUNCTION capacity
 
   FUNCTION temperature (enthalpy)
     IMPLICIT NONE
     REAL (pr), INTENT(IN) :: enthalpy(:)
     REAL (pr) :: temperature(SIZE(enthalpy))
-    IF (capacity_der.LE.1e-6) THEN
-        temperature = enthalpy_fusion(enthalpy)
+    REAL (pr) :: h1(SIZE(enthalpy)), h2(SIZE(enthalpy))
+    ! IF (temperature_method.EQ.0) THEN
+      IF (capacity_der_solid.LE.1e-6) THEN
+          temperature = enthalpy_fusion(enthalpy)
+     !  ELSE
+     !    ! There are some issues that we need to calculate temperature_S,L first cause here in model we have only
+     !   ! WHERE (enthalpy.LE.enthalpy_S)
+     !      temperature = (SQRT(1.0_pr + 2*capacity_der_solid*(enthalpy_fusion(enthalpy))) - 1.0_pr)/capacity_der_solid
+     !   ! ELSEWHERE (enthalpy_S.LT.enthalpy .AND. enthalpy.LT.enthalpy_L)
+     !   !   a = (conductivity_der_solid + conductivity_der_liquid)/2 +
+     !   !    temperature = (SQRT(1.0_pr + 2*capacity_der_solid*(enthalpy_fusion(enthalpy_S))) - 1.0_pr)/capacity_der_solid +
+     ! END IF
     ELSE
-        !temperature = enthalpy_fusion(enthalpy)
-        temperature = (SQRT(1.0_pr + 4*capacity_der*enthalpy_fusion(enthalpy)) - 1.0_pr)/2/capacity_der
+        h1 = 1.0_pr + (capacity_der_solid - capacity_der_liquid + capacity_fusion)*liquid_fraction(enthalpy)
+        h2 = capacity_der_solid*(1.0_pr - liquid_fraction(enthalpy)) + capacity_der_liquid*liquid_fraction(enthalpy)
```

Comment:
It is easily to use only 1 call, but I suggest to get `liquid_fraction` as parameter. In this case, you never think about optimization.